[Generator] Add FastAPI server for inference engine client

CharlieFRuan opened the pull request:
This is a first stab for https://github.com/NovaSky-AI/SkyRL/issues/96

We merge this PR to a separate branch `http-inf_engine-openai` for now for users to build on top of.

The intended use case is when users have their existing agent harness and want to do rollout with the existing code while using SkyRL.

After setting the following configs:
```
generator:
  use_http_server_inference_engine_client: true
  http_server_inference_engine_client_host: "127.0.0.1"
  http_server_inference_engine_client_port: 8000
```
user can post OpenAI requests to `127.0.0.1:8000` in tools like `openai.OpenAI` or `litellm`. You can refer to `skyrl-train/examples/rollout_with_http_server_litellm.py` on how to use this feature.

Underlying-ly, when these configs are set, we launch a FastAPI server when calling `InferenceEngineClient.__init__()`, and only return when the server is ready.

We support OpenAI API, including features like `response_format`, but not features like `stream` or `tools` yet. You can see the supported / unsupported fields in `openai_api_protocol.py`.

### This PR is verified with:
- Running `run_gsm8k.sh` with `skyrl_gym_generator.py` but with `generate_with_http_server()` rather than `inference_engine_client.generate()`. Note this `generate_with_http_server()` is only intended for testing
- Unit tests `tests/gpu/test_http_inference_engine_client.py` and `skyrl-train/tests/cpu/http/test_openai_request_utils.py`
- `rollout_with_http_server_litellm.py` to check compatibility with tools like LiteLLM

However, how to balance such ease of use with token-in-token-out (https://github.com/NovaSky-AI/SkyRL/issues/123) is still under discussion. 

### TODOs:
- [x] Clean up `openai_api_protocol.py` -- be clear about the fields we do not support right now
- [x] CPU unit tests for `convert_inference_output_to_openai()` and `convert_openai_to_inference_input()`
- [x] Add a config flag that will spin up this server; test with `skyrl_gym_generator.py` E2E on `run_gsm8k.sh`
- [x] Address `trajectory_id`
- [x] Be more pragmatic about translating ChatCompletionRequest' sampling-related params to our `sampling_params` in `launch_inference_engine_http_server.py`
- [x] Check error handling
- [x] Test E2E with SkyRLGymGenerator with HTTP server, make sure performance not affected
  - Before 0.5B run_gsm8k takes 100sec for each step's generate, now it takes 130sec. It should be the performance difference between adding each request to the vLLM engine (with HTTP server) vs. batched generate with an offline engine.

### Quick note:
- If we use `OpenAI.chat.completions.create()` in `skyrl_gym_generator.py` without setting a maximum limit on the number of HTTP requests, it will hang. Using `aiohttp.ClientSession().post` does not have this issue. Revisit once this issue is encountered again.

tyler-griggs commented on skyrl-train/skyrl_train/inference_engines/inference_engine_client.py:
```diff
@@ -15,8 +15,10 @@ class InferenceEngineClient(InferenceEngineInterface):
-    def __init__(self, engines: List[InferenceEngineInterface]):
+    def __init__(self, engines: List[InferenceEngineInterface], generator_config: Dict[str, Any]):
         self.engines = engines
+        self.generator_config = generator_config
```
qq: is this used?

SumanthRH replied:
Yeah I think having it be fully OpenAI compatible is a good first step, before introducing such overrides

tyler-griggs commented on skyrl-train/skyrl_train/inference_engines/inference_engine_client.py:
```diff
@@ -15,8 +15,10 @@ class InferenceEngineClient(InferenceEngineInterface):
+    def __init__(self, engines: List[InferenceEngineInterface], generator_config: Dict[str, Any]):
         self.engines = engines
+        self.generator_config = generator_config
+        self.model_name = generator_config.get("model_name")
```
qq: is this only used in tests? If yes, is it actually needed here?

tyler-griggs commented on skyrl-train/skyrl_train/inference_engines/launch_inference_engine_http_server.py:
```diff
@@ -0,0 +1,272 @@
+    _global_backend = backend
+
+
+def convert_openai_to_inference_input(request: ChatCompletionRequest, backend: str) -> InferenceEngineInput:
```
Based on your work here, do you think we should update the InferenceEngineInput/Output types to better match ChatCompletionRequest/Response?

tyler-griggs commented on skyrl-train/skyrl_train/inference_engines/openai_api_protocol.py:
```diff
@@ -0,0 +1,139 @@
+    content: str
+
+
+class ChatCompletionRequest(BaseModel):
```
Is there a reason not to directly use `openai.types.chat` to get these types? Is it too heavy-weight for some reason?

tyler-griggs reviewed (commented):
I left a handful of questions in this first pass. My primary question is about the intended user experience of creating a custom generator. Ideally, a user just flips a config flag or in the custom generator calls into a single method (probably exposed by the inference engine client), and receives the URL to an OpenAI API compatible endpoint. The current example code in the `skyrl_gym_generator.py` is more than we'd like to make a user implement.

CharlieFRuan commented on skyrl-train/skyrl_train/generators/skyrl_gym_generator.py:
```diff
@@ -34,10 +41,46 @@ def __init__(
+                max_wait_seconds=30,
+            )
+            # Store the base URL for direct HTTP requests
+            self.base_url = f"http://{self.http_server_inference_engine_client_host}:{self.http_server_inference_engine_client_port}"
```
Good point, moved the spinning up / shutting down into `InferenceEngineClient`, triggered based on generator config.

CharlieFRuan commented on skyrl-train/skyrl_train/generators/skyrl_gym_generator.py:
```diff
@@ -49,6 +92,55 @@ def __init__(
+            except Exception:
+                pass
+
+    async def _openai_generate(
```
> don't have to re-implement this

Agreed, moved to a static helper method `generate_with_http_server()`.

> sit inside the the inference engine client

If users are calling `inference_engine_client.xxx()`, they might as well just not use the HTTP server right...

CharlieFRuan commented on skyrl-train/skyrl_train/inference_engines/inference_engine_client.py:
```diff
@@ -15,10 +16,79 @@ class InferenceEngineClient(InferenceEngineInterface):
+            except Exception as e:
+                print(f"Error shutting down HTTP server: {e}")
+
+    def __getstate__(self):
+        """
+        Override to avoid pickling the server thread.
+        Needed when passing InferenceEngineClient as an argument to async_run_ray_method().
+        """
+        state = self.__dict__.copy()
+        # Thread objects are not picklable – just drop the reference.
+        state["_server_thread"] = None
+        return state
```
Not sure if this is good practice, but needed if we want to spin up server inside inference engine client. Alternative is to spin up server in `main_base.py`

SumanthRH replied:
Yes this is fine. Server thread will be alive as long as the original inference engine client object is alive in the training loop, ray actors can freely interact with it.

CharlieFRuan commented on skyrl-train/skyrl_train/inference_engines/inference_engine_client.py:
```diff
@@ -15,10 +16,79 @@ class InferenceEngineClient(InferenceEngineInterface):
+            f"InferenceEngineClient HTTP server started on {self.http_server_inference_engine_client_host}:{self.http_server_inference_engine_client_port}"
+        )
+
+    def __del__(self):
+        """
+        Destructor to shut down the HTTP server if it was started.
+        """
+        if self.use_http_server_inference_engine_client:
+            try:
+                from skyrl_train.inference_engines.launch_inference_engine_http_server import shutdown_server
+
+                shutdown_server(
+                    host=self.http_server_inference_engine_client_host,
+                    port=self.http_server_inference_engine_client_port,
+                    max_wait_seconds=5,
+                )
+                if hasattr(self, "_server_thread") and self._server_thread.is_alive():
+                    self._server_thread.join(timeout=5)
+            except Exception as e:
+                print(f"Error shutting down HTTP server: {e}")
```
Agreed. We should simply move the code to `teardown()`. Since `_handle_termination` is not called anywhere yet, will leave the `__del__` for now. Graceful shutdown in general is left as a TODO

CharlieFRuan commented on skyrl-train/skyrl_train/generators/skyrl_gym_generator.py:
```diff
@@ -108,7 +126,14 @@ async def agent_loop(
+                    input_batch=engine_input,
+                )
+            else:
+                engine_output = await self.inference_engine_client.generate(engine_input)
```
Addressed.

SumanthRH commented on skyrl-train/examples/rollout_with_http_server_litellm.py:
This is great. Can we actually create a folder called `http_server` in `examples/` ? 

```bash
http_server/
   |- rollout_with_http_server_litellm.py # standalone for showing the feature
   |- run_with_http_server.sh # bash script for E2E training with GSM8K and http server
```

SumanthRH commented on skyrl-train/examples/rollout_with_http_server_litellm.py:
````diff
@@ -0,0 +1,139 @@
+Note that `init_to_simulate_trainer()` is not needed in your custom generator, we
+do it here to simulate what the trainer will instantiate.
+
+Set the following configs in your bash script:
+```
+generator:
+  use_http_server_inference_engine_client: true
+  http_server_inference_engine_client_host: "127.0.0.1"
+  http_server_inference_engine_client_port: 8000
+```
````
This part is best just put in a standalone bash script...it's confusing to mention this here since this script is meant to be standalone

SumanthRH commented on skyrl-train/skyrl_train/inference_engines/launch_inference_engine_http_server.py:
```diff
@@ -0,0 +1,323 @@
+    async def chat_completion(request: ChatCompletionRequest, raw_request: Request):
+        return await handle_chat_completion(request, raw_request)
+
+    # Health check endpoint
+    @app.get("/health")
+    async def health_check():
+        return {"status": "healthy"}
+
```
```suggestion
    # Health check endpoint
    # All inference engine replicas are initialized before creating `InferenceEngineClient`, and thus
    # we can start receiving requests as soon as the FastAPI server starts
    @app.get("/health")
    async def health_check():
        return {"status": "healthy"}

```

SumanthRH replied:
Thought we'd add this since this is a difference from say vllm

SumanthRH commented on skyrl-train/skyrl_train/generators/skyrl_gym_generator.py:
```diff
@@ -34,10 +35,27 @@ def __init__(
+            self.base_url = f"http://{self.http_server_inference_engine_client_host}:{self.http_server_inference_engine_client_port}"
+        else:
+            self.base_url = None
+
```
Is `use_conversation_multi_turn` supported with the http server? I think not, because we'd need a `/completions` endpoint that supports token in, token out. 

Can you raise an error here if `use_conversation_multi_turn=False` with http server?

CharlieFRuan commented:
Thanks for the review; addressed all comments.

GSM8K timing on 1 A100 for 0.5B:
<img width="876" height="542" alt="image" src="https://github.com/user-attachments/assets/b6c74041-5219-4e4d-b78e-a5e523c1ff2a" />
This performance discrepancy I think is expected because we are breaking a previously batched N request into N HTTP requests.

Reward is identicaly:
<img width="1286" height="510" alt="image" src="https://github.com/user-attachments/assets/def5fe69-234d-470d-8c0e-67b26040b6cc" />

Still running Search R1, which should have much less performance difference.

CharlieFRuan commented:
For SearchR1 (max_turn=2, 1xH100, 0.5B model) we have roughly the same time cost for the generate step between HTTP server and `.generate()`.

CharlieFRuan commented:
for future reference, the litellm example needs this fix: https://github.com/NovaSky-AI/SkyRL/pull/153
